rfd/00xx-kubernetes-access-cluster-discovery.md

```diff
@@ -0,0 +1,428 @@
+kind: ClusterRole
+metadata:
+  name: teleport-role
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - users
+  - groups
+  - serviceaccounts
+  verbs:
+  - impersonate
+- apiGroups:
+  - ""
+  resources:
+  - pods
+  verbs:
+  - get
+- apiGroups:
+  - "authorization.k8s.io"
+  resources:
+  - selfsubjectaccessreviews
+  - selfsubjectrulesreviews
+  verbs:
+  - create
+```
+
+The second step is to create the cluster role binding to map the cluster role into the Kubernetes Teleport group.
+
+```yaml
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: teleport-role-binding
+subjects:
+- kind: Group
+  name: system:teleport
+  apiGroup: rbac.authorization.k8s.io
+```
+
+Finally, the operator has to map the Teleport IAM role into the Kubernetes group in the `configmap/aws-auth`:
+
+```yaml
+apiVersion: v1
+data:
+  mapRoles: |
+    - groups:
+      - system:teleport
+      rolearn: arn:aws:iam::222222222222:role/teleport-role
+      username: system:teleport
+...
+```
+
+At this point, Teleport discovery can run the cluster discovery mechanism every minute and monitor if new clusters appeared or if any was deleted. When a new cluster is found, Teleport checks if it has access to it by generating short-lived credentials and doing requests to the API. If the request is successful, it registers the cluster by creating the cluster config and stores the short-lived credentials in the credentials map. Teleport will also import, for each discovered cluster, the EKS cluster tags as labels.
+
+If the request fails due to forbidden access, Teleports logs an error mentioning that the IAM role map process is not working correctly.
+
+For security reasons, instead of creating a new service account that provides a static token, like the one used by the Kubeconfig method, which can be stolen, the credentials used are the ones provided by [AWS IAM authenticator][awsiamauthenticator] since they are only valid for short periods, the period can be controlled by the operator at IAM role level, and Teleport can rotate them once they are near expiration.
+
+If a cluster is deleted or no longer matches the discovery criteria, the credentials are wiped, and the cluster entry is removed.
+
+With this method, Teleport does not install any agent in the cluster, but it requires that the Kubernetes API is accessible to the Teleport discovery. This means that if the operator wants to enroll clusters with private API endpoints, he must configure the Teleport discovery in a machine that has the ability to access the endpoint.
+
+
+### Limitations
+
+Teleport will only provide access to API and will not enroll, automatically, databases or applications like Prometheus or Grafana that easily configured in a situation where Teleport Agent is installed into the cluster.
+
+The IAM mapping between Teleport IAM Role and Kubernetes roles is a complex and tedious process that must be done by the operator. Without it, Teleport cannot enroll the cluster.
```

Comment:
I would link this issue here: https://github.com/aws/containers-roadmap/issues/185.

Seems like many people are running into this problem (unsurprisingly) and the EKS team might be working on the API to automate this.